skimage/morphology/max_tree.py

```diff
@@ -0,0 +1,667 @@
+           DOI:10.1007/3-540-36104-9_23
+    .. [2] Carlinet, E., & Geraud, T. (2014). A Comparative Review of
+           Component Tree Computation Algorithms. IEEE Transactions on Image
+           Processing, 23(9), 3885-3895.
+           DOI:10.1109/TIP.2014.2336551
+
+    Examples
+    --------
+    We create an image (quadratic function with a minimum in the center and
+    4 additional local minima.
+
+    >>> w = 12
+    >>> x, y = np.mgrid[0:w,0:w]
+    >>> f = 180 + 0.2*((x - w/2)**2 + (y-w/2)**2)
+    >>> f[2:3,1:5] = 160; f[2:4,9:11] = 140; f[9:11,2:4] = 120
+    >>> f[9:10,9:11] = 100; f[10,10] = 100
+    >>> f = f.astype(np.int)
+
+    We can calculate the diameter closing:
+
+    >>> closed = diameter_closing(f, 3, connectivity=1)
+
+    All small minima with a maximal extension of 2 or less are removed.
+    The remaining minima have all a maximal extension of at least 3.
+
+
+    Note
+    ----
+    If a max-tree representation (parent and tree_traverser) are given to the
+    function, they must be calculated from the inverted image for this
+    function, i.e.:
+    >>> P, S = max_tree(invert(f))
+    >>> closed = diameter_closing(f, 3, parent=P, tree_traverser=S)
+    """
+    # inversion of the input image
+    image_inv = invert(image)
+    output = image_inv.copy()
+
+    if parent is None or tree_traverser is None:
+        parent, tree_traverser = max_tree(image_inv, connectivity)
+
+    diam = _max_tree._compute_extension(image_inv.ravel(),
+                                        np.array(image_inv.shape,
+                                                 dtype=np.int32),
+                                        parent.ravel(), tree_traverser)
+
+    _max_tree._direct_filter(image_inv.ravel(), output.ravel(), parent.ravel(),
+                             tree_traverser, diam, diameter_threshold)
+    output = invert(output)
+    return output
+
+
+def max_tree_local_maxima(image, connectivity=2,
+                          parent=None, tree_traverser=None):
+    """Determine all local maxima of the image.
+
+    The local maxima are defined as connected sets of pixels with equal
+    grey level strictly greater than the grey levels of all pixels in direct
+    neighborhood of the set. The function labels the local maxima.
+
+    Technically, the implementation is based on the max-tree representation
+    of an image. The function is very efficient if the max-tree representation
+    has already been computed. Otherwise, it is preferable to use
+    the function local_maxima.
+
+    Parameters
+    ----------
+    image : ndarray
+        The input image for which the maxima are to be calculated.
+    connectivity: unsigned int, optional
+        The neighborhood connectivity. The integer represents the maximum
+        number of orthogonal steps to reach a neighbor. In 2D, it is 1 for
+        4-connectivity and 2 for 8-connectivity. Default value is 1.
+    parent: ndarray, int64, optional
```

Comment:
Now that I understand these things, I don't know whether I love `parent` as the parameter name. What do people think of:

- tree, parent_indices, tree_image
- traversal_order, tree_traversal_order, topological_sorted_indices, sorted_indices, tree_sorted_indices

?

Reply:
I would keep either `parent or `parent_indices to be conform with the publication. For the second parameter, I do not have a strong opinion. Perhaps `topological_sorted_indices describes this most clearly?

Reply:
Any opinion?